### PR TITLE
Only redefine _WIN32_WINNT macro when < 0x0501

### PR DESCRIPTION
**Pull request by Faless**
## Status
**READY**

## Requires Backporting
Yes (bugfix)

Which branch?
`mbedtls-2.7`, `mbedtls-2.8` (why is it not branched yet?!)

## Migrations
NO

## Additional comments
`net_sockets.c` was redefining the `_WIN32_WINNT` to ensure `getaddrinfo()` and the like to work.
This should only be done when `_WIN32_WINNT < 0x0501` or it might break builds when embedding in projects that require higher `_WIN32_WINNT` minimum version.

## Todos
- [ ] Tests
- [ ] Backported

### CODE REVIEW

**Comment by irwir** [comment]
No such changes for 2.8.0 could be expected: [Retiring Legacy Windows Support](https://tls.mbed.org/tech-updates/blog/retiring-legacy-windows-support)

**Comment by Faless** [comment]
@irwir problems arises when you build the reset of your program with **higher** versions, not lower (e.g. `0x0601`)

**Comment by irwir** [comment]
Most probably this re-definition is not needed.
XP build works fine with the code from PR #1427.

**Comment by Faless** [comment]
Was introduced in 2013-14 by 1321135758435ef60f0d689153632ceb1fce8ac9 and 3b6269aa0852abe03c7364e150a09982f44a865b apparently to fix MinGW problems, I just tried to be conservative, but I guess it can be removed altogether then

**Comment by irwir** [comment]
`Winsock.h` header being caused troubles, it would not compile in VS 2017. Therefore being _conservative_ was not an option.
No real tests with MinGW builds though, only automatic checks passed for the mentioned PR.

**Comment by Faless** [comment]
> Winsock.h header caused troubles, it would not compile in VS 2017. Therefore being conservative was not an option.

I don't see how this is related to my usage of *being conservative*.
The change I proposed has nothing to do with the included header, it's about not overruling a preprocessor macro.

The code in `library/net_sockets.c` does not even include `winsock.h`  (did you mean `winsock2.h`?)

It seems to me you didn't check/understand the scope of this PR.

**Comment by irwir** [comment]
Sorry, my mistake, `winsock.h` was in another repository together with `winsock2.h`.
Anyway, here removing _WIN32_WINNT was a small part of the changes to get the code compiled.

**Comment by Faless** [comment]
@sbutcher-arm thanks, I already signed the CLA (registered with my GitHub email and signed), you can proceed to verification I guess.

I also rebased on the current `development` branch as merging was blocked due to conflicts.

**Comment by RonEld** [comment]
@Faless Could you please send us a link to your Mbed acount , as Faless is not found as a user

**Comment by RonEld** [comment]
@Faless Thank you for sending us the link!
As @irwir mentioned, in #1427 The `_WIN32_WINNT` override was removed altogether, which conflicts with this PR. We will need to decide which is the best approach. We are open to any of your ideas.
To answer your question about backporting branches, `mbedtls-2.8` is not a LTS release, and as such no branch was created. The branches that need backporting are:  `mbedtls-2.1` and `mbedtls-2.7`.

**Comment by Faless** [comment]
> The _WIN32_WINNT override was removed altogether, which conflicts with this PR. We will need to decide which is the best approach.

Well, #1427 takes longer to test, has conflicts in it, and can't be easily backported to older versions.

Of course if/when #1427 is merged my change can go away (I really don't care), but I fail to understand how it is possible that this bug is still here after `2.9`, `2.10`, `2.1.12` and `2.7.3` when a simple fix like this and with no side effects is available.

**Comment by sbutcher-arm** [comment]
Apologies @Faless, that, as you say, a simple easy fix hasn't been merged yet. We only have limited capacity and simply have a lot of PR's to work through, so haven't got to this one yet. 

I've just approved it, and given it's simplicity, I think it makes more sense to get this one in sooner than #1427, which will take more effort, even if ultimately we merge #1427 soon after.

This needs backports which I'll handle as new PR's.

**Comment by sbutcher-arm** [comment]
Added backports for `mbedtls-2.7` and `mbedtls-2.1`. The PR also needs a ChangeLog entry but can be added during merge. Text can be taken from 4e5edf318475 - 'Add ChangeLog entry for _WIN32_WINNT override fix'.

**Comment by Faless** [comment]
> We only have limited capacity and simply have a lot of PR's to work through, so haven't got to this one yet.

No worries, and thanks for the approval.
I'm looking forward to have one less downstream patch to apply :)

**Comment by irwir** [comment]
The scope of this PR should have been limited even further, so that redefinition applies to only those compilers that require such action.

**Comment by Faless** [comment]
> compilers that require such action

@irwir what do you mean with that? That macro is used by windows headers, not compilers. Do you have an example?

**Comment by irwir** [comment]
As your research suggests, only old MinGW had issues.
Then It would be unnecessary for Visual Studio.

**Comment by Faless** [comment]
>  Then It would be unnecessary for Visual Studio.

Visual Studio versions are totally unaffected by the change because they always define a `_WIN32_WINNT >= 0x0501`.

**Comment by akien-mga** [comment]
> As your research suggests, only old MinGW had issues.
Then It would be unnecessary for Visual Studio.

Aren't we bikeshedding here? The code that @Faless is changing was obviously wrong (it was undefining whatever `_WIN32_WINNT` defined by the user of mbedtls to replace it by an arbitrary, potentially older value). The change made fixes that, and that's it. It doesn't make things much clever, and it doesn't have to be made compiler specific either.

The code *as it is today* causes issues for all compilers. @Faless's research just showed that this bad code was introduced to workaround issues with MinGW, but since he's not undo this change, it doesn't have any impact on the fix he's doing (the code fixes MinGW issues, but also impacted other compilers with this forced redefinition).

It seems to me like we're discussing this a lot, but did anyone actually check the diff of this PR? It's pretty obvious that it's a good change and fixes a previously bogus redefinition. Whether it's the *perfect* way to improve this piece of code is out of the scope of this issue. And #1427 already takes cares of this part.

**Comment by irwir** [comment]
> Aren't we bikeshedding here?

We should be discussing quality of the patch without overreaction, shall we not?
The suggested change goes in the right direction, but does not go far enough,

Both the original and your patches were made with disregard to the fact that there are more compilers that  MinGW only.
While a good patch should fix issues without creating new ones.
It would be correct at least to add a check for MinGW.
For the perfect solution add a check for the problematic versions of the compiler.

**Comment by akien-mga** [comment]
@irwir Again, please read the diff and try to understand what this code does.

In plain English:
If `_WIN32_WINNT` is not defined, or if it's defined but lower than `0x0501`, it should be redefined to `0x0501` before including the Windows header which requires this version to define the proper interfaces.

Compilers are not involved here, those headers have similar requirements in MinGW and in the VS SDK. And even if the VS header had somehow removed this `_WIN32_WINNT` minimal requirement and/or would enforce that you are no longer allowed to ever compile for a Windows version below `0x0501`, this check would still be harmless.

Either `_WIN32_WINNT` is used, and then this patch is needed, or it's not used and then this patch is harmless.

As I mentioned, again, since there is already another PR which handles refactoring this code and removing the need for this redefinition, I don't know what you want more from this PR.

Eventually, @Faless and I are downstream users and we don't care what gets merged as long as we don't need to keep maintaining our own patches to be able to use mbedtls. So feel free to make another PR that supersedes this one, and mbedtls maintainers can decide which one they want.

**Comment by irwir** [comment]
> are downstream users and we don't care

That is the key phrase.

Those who merge the changes into widly used library are taking into account many more things.
Please do not try to read my comments as personal attack, and have a nice day.

**Comment by mpg** [comment]
> While a good patch should fix issues without creating new ones.

@irwir Can you name a new issue that this patch would introduce? It looks to me like it is a step in the right direction (as you said) that fixes a real issue that people are having now, so unless it's actually introducing a new issue I don't see a reason not to merge it.

Perhaps there are better ways to fix this, such as (part of) #1427, but none of them is as ready as this patch is, so if this patch is good enough, let's merge it now _and_ have the better fix later, I don't think we have to choose in this case.

**Comment by irwir** [comment]
> Can you name a new issue that this patch would introduce?

It was a general statement about good patches; and the old original patch was not very good.
This change restricts the scope of the original patch, but does that insufficiently.

Old versions of MinGW compilers had issues, therefore it would have been correct to limit the scope or redefinition to MinGW only. It was not done before, why not do this now?
Additional checks are as simple as:
```
#ifdef __GNUC__
#ifdef __MINGW32__
```
Additionally, as mentioned above, the current `development` branch supports only Vista and above, this redefinition became poitless and should be removed completely. Backports only, then.

**Comment by Faless** [comment]
That is incorrect, old versions of MSVC might be affected too if they
define a `_WIN32_WINNT` value less than XP (`0x0501`).
Recent versions are not, but old ones might, and only checking for MinGW
would be conceptually wrong.

**Comment by irwir** [comment]
@ Faless 
> only checking for MinGW would be conceptually wrong.

Is there a known issue with supported versions of MSVC?
Otherwise it would fixing what was not broken for vague `might be` reasons.

**Comment by Faless** [comment]
You clearly do not understand the patch, how the `_WIN32_WINNT` macro
works, and how it changes the behaviour of Windows headers.
I really have no time explaining all this stuff to a non maintainer
(especially since maintainer clearly understood and accepted it).
The redefine bug affected all MSVC versions, the provided fix works
for all MSVC versions, that's it.

On Mon, Jun 11, 2018 at 2:28 PM, irwir <notifications@github.com> wrote:
> @ Faless
>
> only checking for MinGW would be conceptually wrong.
>
> Is there a known issue with supported versions of MSVC?
> Otherwise it would fixing what was not broken for vague might be reasons.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.



-- 
Ciao,
        Fabio

**Comment by sbutcher-arm** [comment]
To respond to some of the comments on this PR - for the moment, I just want to fix @Faless's problem as it's quick and immediate, but longer term we should try and get rid of defining `_WIN32_WINNT` altogether.
